### PR TITLE
Create updateServiceProvider method

### PR DESCRIPTION
**Pull request by hwupathum**
Create updateServiceProvider SOAP and REST methods, and adding new unit tests

### Related Issues
- https://github.com/wso2/product-is/issues/16151

### CODE REVIEW

**Comment by hwupathum** [comment]
Need to be merged after https://github.com/wso2/carbon-identity-framework/pull/4753 and framework bump

**Review by madurangasiriwardena — DISMISSED** [review]

File: components/org.wso2.carbon.identity.sso.saml.stub/src/main/resources/IdentitySAMLSSOConfigService.wsdl [hunk 1]
```diff
@@ -50,6 +50,20 @@
                     </xs:sequence>
                 </xs:complexType>
             </xs:element>
+            <xs:element name="updateRPServiceProvider">
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.sso.saml.stub/src/main/resources/IdentitySAMLSSOConfigService.wsdl:53]: Generate wsdl

File: components/org.wso2.carbon.identity.sso.saml.ui/src/main/java/org/wso2/carbon/identity/sso/saml/ui/client/SAMLSSOConfigServiceClient.java [hunk 2]
```diff
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2010, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ * Copyright (c) 2010, WSO2 LLC. (http://www.wso2.com).
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.sso.saml.ui/src/main/java/org/wso2/carbon/identity/sso/saml/ui/client/SAMLSSOConfigServiceClient.java:2]: (2010-2023)

File: components/org.wso2.carbon.identity.sso.saml.ui/src/main/java/org/wso2/carbon/identity/sso/saml/ui/client/SAMLSSOConfigServiceClient.java [hunk 3]
```diff
@@ -72,6 +72,23 @@ public boolean addServiceProvider(SAMLSSOServiceProviderDTO serviceProviderDTO)
         return status;
     }
 
+    /**
+     * Update the service provider if exists
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.sso.saml.ui/src/main/java/org/wso2/carbon/identity/sso/saml/ui/client/SAMLSSOConfigServiceClient.java:76]: Method comment .

File: components/org.wso2.carbon.identity.sso.saml.ui/src/main/java/org/wso2/carbon/identity/sso/saml/ui/client/SAMLSSOConfigServiceClient.java [hunk 4]
```diff
@@ -72,6 +72,23 @@ public boolean addServiceProvider(SAMLSSOServiceProviderDTO serviceProviderDTO)
...
+     *
+     * @param serviceProviderDTO service provider DTO
+     * @return true if the update is successful
+     * @throws IdentitySAML2SSOUiException if an error occurs while updating the service provider
+     */
+    public boolean updateServiceProvider(SAMLSSOServiceProviderDTO serviceProviderDTO) throws IdentitySAML2SSOUiException {
+        boolean status  ;
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.sso.saml.ui/src/main/java/org/wso2/carbon/identity/sso/saml/ui/client/SAMLSSOConfigServiceClient.java:83]: fix formatting

File: components/org.wso2.carbon.identity.sso.saml.ui/src/main/java/org/wso2/carbon/identity/sso/saml/ui/client/SAMLSSOConfigServiceClient.java [hunk 5]
```diff
@@ -72,6 +72,23 @@ public boolean addServiceProvider(SAMLSSOServiceProviderDTO serviceProviderDTO)
...
+        try {
+            status = stub.updateRPServiceProvider(serviceProviderDTO);
+        } catch (RemoteException | IdentitySAMLSSOConfigServiceIdentityException e) {
+            throw new IdentitySAML2SSOUiException("Error while uploading the service provider", e);
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.sso.saml.ui/src/main/java/org/wso2/carbon/identity/sso/saml/ui/client/SAMLSSOConfigServiceClient.java:87]: error while updating

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOConfigService.java [hunk 6]
```diff
@@ -46,6 +46,22 @@ public boolean addRPServiceProvider(SAMLSSOServiceProviderDTO spDto) throws Iden
...
     }
 
+    /**
+     * Updates SAML Service provider information.
+     *
+     * @param spDto SAMLSSOServiceProviderDTO containing service provider configurations.
+     * @return true if the update is successful.
+     * @throws IdentityException if an error occurs while updating the service provider.
+     */
+    public boolean updateRPServiceProvider(SAMLSSOServiceProviderDTO spDto) throws IdentityException {
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOConfigService.java:56]: rename `spDto`

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOConfigServiceImpl.java [hunk 7]
```diff
@@ -88,6 +89,23 @@ public boolean addRPServiceProvider(SAMLSSOServiceProviderDTO spDto) throws Iden
...
     }
 
+    /**
+     * Updates a SAML service provider.
+     *
+     * @param spDto DTO containing the SAML SP configuration.
+     * @return true if the update was successful.
+     * @throws IdentityException if an error occurs while updating the SAML SP.
+     */
+    public boolean updateRPServiceProvider(SAMLSSOServiceProviderDTO spDto) throws IdentityException {
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOConfigServiceImpl.java:99]: rename `spDto`

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOConfigServiceImpl.java [hunk 8]
```diff
@@ -126,6 +144,44 @@ public SAMLSSOServiceProviderDTO createServiceProvider(SAMLSSOServiceProviderDTO
...
+     * Update a SAML service provider.
+     *
+     * @param spDto DTO containing the SAML SP configuration.
+     * @return SAMLSSOServiceProviderDTO with the information on the SAML SP.
+     * @throws IdentityException if an error occurs while updating the SAML SP.
+     */
+    public SAMLSSOServiceProviderDTO updateServiceProvider(SAMLSSOServiceProviderDTO spDto) throws IdentityException {
+
+        try {
+            SAMLSSOConfigAdmin configAdmin = new SAMLSSOConfigAdmin(getConfigSystemRegistry());
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOConfigServiceImpl.java:157]: refactor validation part

Comment (hwupathum) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOConfigServiceImpl.java:157]: check if can be removed from try block

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOConfigServiceImpl.java [hunk 9]
```diff
@@ -146,6 +202,28 @@ public SAMLSSOServiceProviderDTO uploadRPServiceProvider(String metadata) throws
...
+            throws IdentitySAML2SSOException {
+
+        try {
+            SAMLSSOConfigAdmin configAdmin = new SAMLSSOConfigAdmin(getConfigSystemRegistry());
+            if (log.isDebugEnabled()) {
+                log.debug("Updating SAML Service Provider with metadata: " + metadata);
+            }
+            return configAdmin.updateRelyingPartyServiceProviderWithMetadata(metadata);
+        } catch (IdentityException e) {
+            String tenantDomain = getTenantDomain();
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOConfigServiceImpl.java:222]: make single line

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOConfigServiceImpl.java [hunk 10]
```diff
@@ -175,6 +253,35 @@ public SAMLSSOServiceProviderDTO createServiceProviderWithMetadataURL(String met
...
+     * Update a service provider with configurations provided via a metadata URL.
+     *
+     * @param metadataUrl URL to fetch the SAML SP metadata file.
+     * @return SAMLSSOServiceProviderDTO with the information on the created SAML SP.
+     * @throws IdentitySAML2SSOException
+     */
+    public SAMLSSOServiceProviderDTO updateServiceProviderWithMetadataURL(String metadataUrl)
+            throws IdentitySAML2SSOException {
+
+        InputStream in = null;
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOConfigServiceImpl.java:266]: rename variables

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOConfigServiceImpl.java [hunk 11]
```diff
@@ -175,6 +253,35 @@ public SAMLSSOServiceProviderDTO createServiceProviderWithMetadataURL(String met
...
+        try {
+            URL url = new URL(metadataUrl);
+            URLConnection con = url.openConnection();
+            con.setConnectTimeout(getConnectionTimeoutInMillis());
+            con.setReadTimeout(getReadTimeoutInMillis());
+            in = new BoundedInputStream(con.getInputStream(), getMaxSizeInBytes());
+
+            String metadata = IOUtils.toString(in);
+            return updateRPServiceProviderWithMetadata(metadata);
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOConfigServiceImpl.java:275]: close connection

Comment (hwupathum) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/SAMLSSOConfigServiceImpl.java:275]: change in create mtd as well

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/admin/SAMLSSOConfigAdmin.java [hunk 12]
```diff
@@ -94,6 +94,36 @@ public boolean addRelyingPartyServiceProvider(SAMLSSOServiceProviderDTO serviceP
         }
     }
 
+    /**
+     * Update a service provider
+     *
+     * @param serviceProviderDTO service Provider DTO
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/admin/SAMLSSOConfigAdmin.java:100]: Sentence case

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/admin/SAMLSSOConfigAdmin.java [hunk 13]
```diff
@@ -128,11 +188,17 @@ private String getIssuerWithQualifier(SAMLSSOServiceProviderDO serviceProviderDO
         return SAMLSSOUtil.getIssuerWithQualifier(serviceProviderDO.getIssuer(), serviceProviderDO.getIssuerQualifier());
     }
 
-    private SAMLSSOServiceProviderDTO persistSAMLServiceProvider(SAMLSSOServiceProviderDO samlssoServiceProviderDO)
-            throws IdentityException {
+    private SAMLSSOServiceProviderDTO persistSAMLServiceProvider(SAMLSSOServiceProviderDO samlssoServiceProviderDO,
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/admin/SAMLSSOConfigAdmin.java:191]: change parameter order

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/admin/SAMLSSOConfigAdmin.java [hunk 14]
```diff
@@ -128,11 +188,17 @@ private String getIssuerWithQualifier(SAMLSSOServiceProviderDO serviceProviderDO
...
+                                                                 boolean isEditingSP) throws IdentityException {
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/admin/SAMLSSOConfigAdmin.java:192]:
```suggestion
                                                                 boolean isUpdateOperation) throws IdentityException {
```

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/admin/SAMLSSOConfigAdmin.java [hunk 15]
```diff
@@ -216,7 +282,41 @@ public SAMLSSOServiceProviderDTO uploadRelyingPartyServiceProvider(String metada
             }
         }
 
-        return persistSAMLServiceProvider(samlssoServiceProviderDO);
+        return persistSAMLServiceProvider(samlssoServiceProviderDO, false);
+    }
+
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/admin/SAMLSSOConfigAdmin.java:287]: remove extra line

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/admin/SAMLSSOConfigAdmin.java [hunk 16]
```diff
@@ -216,7 +282,41 @@ public SAMLSSOServiceProviderDTO uploadRelyingPartyServiceProvider(String metada
...
+     * @throws IdentityException If an error occurs while updating the service provider.
+     */
+    public SAMLSSOServiceProviderDTO updateRelyingPartyServiceProviderWithMetadata(String metadata)
+            throws IdentityException {
+
+        Parser parser = new Parser(registry);
+        SAMLSSOServiceProviderDO samlssoServiceProviderDO = new SAMLSSOServiceProviderDO();
+
+        try {
+            //pass metadata to samlSSOServiceProvider object
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/admin/SAMLSSOConfigAdmin.java:303]: comments sentence case

**Comment by hwupathum** [comment]
Commented couple of lines in the generated wsdl due to https://github.com/wso2/product-is/issues/16227

**Review by mevan-karu — APPROVED** [review]

**Pull request merged**